### ISSUE TITLE
Add modifier functions

### ISSUE THREAD

**Issue by AlexanderJDupree** [AlexanderJDupree/LinkedListsCPP#10]
- [x] push_front() adds element to the beginning of the list
Ex.

```c++
LinkedList<int> list;

list.push_front(5);
list.push_front(4);

assert(*list.begin() == 4);
```
This assertion will pass because 4 was pushed to the front of the list

- [x] push_back() adds element to the back of the list
Ex.

```c++
LinkedList<int> list;
list.push_back(5);
list.push_back(4);

assert(*list.begin() == 5);
```
This assertion will pass because 4 was pushed to the back of the last after 5

- [x] insert() inserts an element before a specified position 
Ex.

```c++
LinkedList<int> list = { 1, 2, 3, 4, 5};
LinkedList<int>::iterator it = list.begin();

++it // it now points to 2

/* Insert a SINGLE ELEMENT */
list.insert(it, 10); // 1, 2, 10, 3, 4, 5
// it still points to 2

/* Insert MULTIPLE ELEMENTS */
list.insert(it,2, 20); // 1, 2, 20, 20, 10, 3, 4, 5 
// it still points to 2

/* Insert a RANGE OF ELEMENTS */
std::vector<int> nums (2, 100);
list.insert(it, nums.begin(), nums.end()) // 1, 2, 100, 100, 20, 20, 10, 3, 4, 5
```
- [x] pop_front() erase the head of the list, with an optional parameter to return the erased value to
Ex.

```c++
LinkedList<int> list = {0, 5, 10, 15, 20};
list.pop_front();   // 5, 10, 15, 20

int value;
list.pop_front(value);  // 10, 15, 20
std::cout << value; // 5
```

- [x] pop_back() erase the tail of the list, with an optional parameter to return the erased value to
Ex.

```c++
LinkedList<int> list = {0, 3, 6, 9, 12, 15};
list.pop_back();  // 0, 3, 6, 9, 12

int value;
list.pop_back(value);  // 0, 3, 6, 9
std::cout << value; // 12
```

- [x] erase() erase a specified element. Returns (and updates iterator argument) an updated iterator of next undeleted element
Ex
```c++ 
LinkedList<int> list = {1, 2, 3, 4, 5};
LinkedList<int>::iterator it1 = list.begin();
++it1 // Points to 2
/* Erase SINGLE ELEMENT */
list.erase(it1);  // 1 3 4 5
// it1 points to 3

LinkedList<int>::iterator it2;
it2 = list.erase(it1);  // 1 4 5
// it1 and it2 both point to 4

list.push_back(6);
list.push_back(7);
list.push_back(8);  // 1 4 5 6 7 8

it2 += 3;  // it2 points to 7

/* Erase RANGE OF ELEMENTS */
list.erase(it1, it2); // 1 7 8
// it1 and it2 both point to 7

// It and beginIt both point to 4
```

- [x] clear() clears the content of the entire list

**Comment by AlexanderJDupree** [comment]
push_front() implemented on commit 14c8029223ea1f842b0fe8a889ad6ebd1b9202ce

**Comment by jake-bickle** [comment]
push_back() implemented on merge #22

**Comment by AlexanderJDupree** [comment]
@jake-bickle implemented insert() and clear() in PR #27

**Comment by AlexanderJDupree** [comment]
Correction, insert() still needs its overloads for it to be complete

**Comment by jake-bickle** [comment]
Erase implemented in PR #34

**Comment by jake-bickle** [comment]
I've decided that in the event that someone is attempting to delete something from an empty list, I'll just throw an error rather than doing nothing. I'll have to update the functions accordingly.

**Comment by AlexanderJDupree** [comment]
I would disagree with the error throwing. As of right now, our class is exception safe, which is STL compliant. If our container starts throwing exceptions we will need to educate clients of our container the what, when, where, and why exceptions are thrown. Whereas, doing nothing is a perfectly acceptable function behavior. 

Take for example this code

```c++
template <typename T>
void clearFirst(LinkedList<T>& list)
{
    list.erase(list.begin());
    return;
}
```
Obviously, this function is pretty useless but lets say the client had need of a subroutine that just takes any list it receives and erases the first element. If we change the code to throw an exception the implementation changes to this.

```c++
template <typename T>
void clearFirst(LinkedList<int>& list)
{
    try
    {
        list.erase(list.begin());
    }
    catch(std::exception& err)
   {
        // Do something. But more than likely, do nothing. which is exactly what the function originally does
    }
   return;
}
```
As you can see the exception throwing implementation has a lot more noise to deal with.

**Comment by jake-bickle** [comment]
Okay, that makes sense. The reason I felt this way is that popping front or back returns the value of the data. However, if the list is empty, there is nothing to return. As this is a very "pythonic" implementation, I decided to go the route that Python did, which is to thrown an error.

Which leads me to my new question then, what should I return if the List is empty?

**Comment by AlexanderJDupree** [comment]
Hmmm . . . . Good question. Yeah in the case of pop front and back it might be appropriate to throw. Maybe that's why the STL containers pop_front and pop_back methods are void. If we can somehow manage to get the function to accomplish nothing if the list is empty that'd be the best case scenario

**Comment by jake-bickle** [comment]
Well we can't simply return nothing. We could return a NULL value if the container holds primitive types, but not if it the container holds an abstract data type. It seems the only options we have is to declare it void or thrown an error.

**Comment by AlexanderJDupree** [comment]
How about this, we create two overloads for the pop_front and pop_back functions. 
This first method, will be a void function that basically just calls erase on the front or back of the list. 
The second overload will take allow the client to pass in an argument of type T, we then pop the data from the back or front of the list into the passed in parameter, then return that parameter.

**Comment by AlexanderJDupree** [comment]
So like this:

```c++
LinkedList<int> list { 1, 2, 3, 4, 5 }
list.pop_front(); // 2, 3, 4, 5;

int x;
list.pop_back(x); // 2, 3, 4

assert(x == 5);
// x equals 5 now. 
```

**Comment by jake-bickle** [comment]
Okay, that can work. I can add this functionality to erase as well.

**Comment by AlexanderJDupree** [comment]
Sounds good! I look forward to reviewing the Pull Request

**Comment by AlexanderJDupree** [comment]
Maybe you can do me a favor and look over commit 20c560ec2dbaa0fd8600f7646f738d8aa10572ed when you get a chance. I added the range based constructed using templated iterators, and it works locally on my machine. However, it fails the Travis CI build due to a seg fault?? When you get a chance checkout the range_constructor branch and try to build and run it.

**Comment by AlexanderJDupree** [comment]
Correction, the commit is 10c2be743307ccc11f08cab5df6a3c15957af2b0

**Comment by AlexanderJDupree** [comment]
Insert modifier finished in PR #42

**Comment by jake-bickle** [comment]
pop_front and pop_back implemented in PR #47
